### PR TITLE
C4-5 Revision History v1

### PR DESCRIPTION
**Pull request by willronchetti**
- Implements `/resource/@@revision-history`, returning all revisions of the item to any user who has edit permissions on that item

### CODE REVIEW

File: snovault/tests/test_storage.py [hunk 1]
```diff
@@ -143,12 +144,14 @@ def test_purge_uuid(session, storage):
     assert current.sid == propsheet.sid
     assert current.rid == resource.rid
 
+    assert len(storage.revision_history(uuid=str(resource.rid))) > 0
```
Comment (netsettler) [inline snovault/tests/test_storage.py:147]: If you're contrasting with the test 7 lines hence, I'd write this test as (untested):
```suggestion
    assert storage.revision_history(uuid=str(resource.rid)) != []
```
Alternatively (xor), you could change the other test to be `== 0`, I suppose. But I think this is good enough.

File: snovault/tests/test_storage.py [hunk 2]
```diff
@@ -178,12 +181,14 @@ def test_delete_compound(session, storage):
     current = session.query(CurrentPropertySheet).one()
     assert current.sid
 
+    assert len(storage.revision_history(uuid=str(resource.rid))) > 0
```
Comment (netsettler) [inline snovault/tests/test_storage.py:184]: Again (untested):
```suggestion
    assert storage.revision_history(uuid=str(resource.rid)) != []
```

File: snovault/tests/test_post_put_patch.py [hunk 3]
```diff
@@ -116,6 +117,9 @@ def test_patch(content, testapp):
     assert res.json['@graph'][0]['simple1'] == 'supplied simple1'
     assert res.json['@graph'][0]['simple2'] == 'supplied simple2'
 
+    revisions = testapp.get(url + '/@@revision-history').json['revisions']
+    assert len(revisions) == 4
+
```
Comment (netsettler) [inline snovault/tests/test_post_put_patch.py:122]: I might prefer a style that did:

* In the block before patching, do:

```
def url_revision_count(url):
    return len(testapp.get(url + '/@@revision-history').json['revisions'])
rh_len = url_revision_count(url)
```

* After each `res = testapp.patch_json` do:

```
rh_len += 1
assert url_revision_count(url) == rh_len
```

* Then you shouldn't need the final two lines here, which are redundant with the suggested alternative test. Also, though, you are immune to adding additional patch requests as long as the count is maintained, as opposed to here you have wired in a 4 which makes the test fragile.

* Optional extra: This `url_revision_count` function could be at toplevel of this testing file and used by other functions.

File: snovault/tests/test_post_put_patch.py [hunk 4]
```diff
@@ -225,12 +229,22 @@ def test_patch_new_schema_version(content, root, testapp, monkeypatch):
...
+        Also tests that links show up in revision history.
+    """
     res = testapp.post_json(COLLECTION_URL, item_with_link[0], status=201)
     url = res.location
 
     testapp.put_json(url, item_with_link[0], status=200)
     testapp.put_json(url, item_with_link[1], status=200)
+    revisions = testapp.get(url + '/@@revision-history').json['revisions']
+    for target_uuid, revision in zip([
+        '775795d3-4410-4114-836b-8eeecf1d0c2f',  # initial post
```
Comment (netsettler) [inline snovault/tests/test_post_put_patch.py:243]: These UUIDs should be extracted from `item_with_link`, not separately wired in here.

**Review by netsettler — APPROVED** [review]
I don't see any blocking issues here.

File: snovault/storage.py [hunk 5]
```diff
@@ -319,6 +319,12 @@ def find_uuids_linked_to_item(self, uuid):
 
         return self.write.find_uuids_linked_to_item(uuid)
 
+    def revision_history(self, uuid):
+        """
+        Gets the revision history for the given uuid from postgres.
+        """
+        return self.write.revision_history(rid=uuid)
```
[the remark this answers is not in the document]
Comment (willronchetti) [reply snovault/storage.py:326]: Yes it is a snovault specific abstraction representing the database. If you were to swap the DB out for something else, nothing would make sense/work, so although its intended to be generic "write" may as well == RDS and "read" == ES.

File: snovault/storage.py [hunk 6]
```diff
@@ -917,7 +932,4 @@ def get_by_username(cls, email):
 
     @classmethod
     def check_password(cls, email, password):
-        user = cls.get_by_username(email)
-        if not user:
-            return False
-        return crypt.check(user.password, password)
+        raise NotImplementedError('This should never happen as we do not manage passwords.')
```
[the remark this answers is not in the document]
Comment (willronchetti) [reply snovault/storage.py:935]: Nothing changed just making sure this is never called and if it is we immediately know.

**Pull request merged**